packages/router/src/params.tsx

```diff
@@ -1,18 +1,41 @@
 import React, { useState, useContext } from 'react'
 
-import { createNamedContext } from './internal'
+import { useLocation } from 'src/location'
+import { useRouterState } from 'src/router-context'
+
+import { createNamedContext, matchPath, parseSearch } from './internal'
 
 export interface ParamsContextProps {
-  params: Record<string, string> | undefined
-  setParams: (newParams: Record<string, string> | undefined) => void
+  params: Record<string, string>
+  setParams: (newParams: Record<string, string>) => void
 }
 
 export const ParamsContext = createNamedContext<ParamsContextProps>('Params')
 
 export const ParamsProvider: React.FC = ({ children }) => {
-  const [params, setParams] = useState<Record<string, string> | undefined>(
-    undefined
-  )
+  const { routes, paramTypes } = useRouterState()
```

Comment:
This is a design question/decision I faced the other day, and I still don't know what the right answer is. Should you use the hook, or should you pass stuff in as props? The provider for this context is defined in a place where we have both `paramTypes` and a flat representation of all routes, so we could just pass them in as props.

I opted to use the hook, just like you (see `<RouteScanner>`). But I have to well grounded reason for doing so...

Reply:
Yeah it's a tough one. I think passing it in is the most flexible since then it's decoupled, but since I need "routes" then I think this is appropriate too.